src/librustc_typeck/check/mod.rs

```diff
@@ -1729,6 +1729,8 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                param_env: ty::ParamEnv<'tcx>,
                body_id: ast::NodeId)
                -> FnCtxt<'a, 'gcx, 'tcx> {
+        FnCtxt::set_emit_default_impl_candidates(inh, body_id);
```

Comment:
What is the goal of *this*? I guess it's meant to allow the `default impl` to "assume itself", as it were? But I think we should handle that instead by having the environment for a `default impl` include itself. In other words:

```rust
default impl<T: Debug> Foo for T {
    fn bar() {
        // Environment would include not only `T: Debug` but also `T: Foo`
    }
}
```

This version promises a bit too much, from what I can tell. For example, I think that this would type-check:

```rust
trait Foo { }
default impl<T> Foo for T { }

trait Bar { fn bar(); }
default impl<T> Bar for T { 
    fn bar() {
        T::foo(); // only provided through a `default impl`
    }
}
```